litex/soc/cores/i2s.py

```diff
@@ -0,0 +1,435 @@
+        sampling), but at 100MHz things will be quite comfortable.
+        
+        The upside of the fully asynchronous implementation is that we can leave the I/O unconstrained,
+        giving the place/route more latitude to do its job.
+        
+        Here's the timing format targeted by this I2S interface:
+        
+            .. wavedrom::
+                :caption: Timing format of the I2S interface
+
+                { "signal" : [
+                  { "name": "clk",         "wave": "n....|.......|......" },
+                  { "name": "sync",        "wave": "1.0..|....1..|....0." },
+                  { "name": "tx/rx",       "wave": ".====|==x.===|==x.=x", "data": ["L15", "L14", "...", "L1", "L0", "R15", "R14", "...", "R1", "R0", "L15"] },
+                ]}
+        
+        - Data is updated on the falling edge
+        - Data is sampled on the rising edge
+        - Words are MSB-to-LSB, left-justified (**NOTE: this is a deviation from strict I2S, which offsets by 1 from the left**)
+        - Sync is an input (FPGA is slave, codec is master): low => left channel, high => right channel
+        - Sync can be longer than the wordlen, extra bits are just ignored
+        - Tx is data to the codec (SDI pin on LM49352)
+        - Rx is data from the codec (SDO pin on LM49352)
+        
+        """)
+
+        # one cache line is 8 32-bit words, need to always have enough space for one line or else nothing works
+        if fifodepth > 504:
+            fifodepth = 504
+            print("I2S warning: fifo depth greater than 504 selected; truncating to 504")
+        if fifodepth < 8:
+            fifodepth = 8
+            print("I2S warning: fifo depth less than 8 selected; truncating to 8")
+
+        # connect pins, synchronizers, and edge detectors
+        if hasattr(pads, 'tx'):
+            tx_pin = Signal()
+            self.comb += pads.tx.eq(tx_pin)
+        if hasattr(pads, 'rx'):
+            rx_pin = Signal()
+            self.specials += MultiReg(pads.rx, rx_pin)
+        sync_pin = Signal()
+        self.specials += MultiReg(pads.sync, sync_pin)
+
+        clk_pin = Signal()
+        self.specials += MultiReg(pads.clk, clk_pin)
+        clk_d = Signal()
+        self.sync += clk_d.eq(clk_pin)
+        rising_edge = Signal()
+        falling_edge = Signal()
+        self.comb += [rising_edge.eq(clk_pin & ~clk_d), falling_edge.eq(~clk_pin & clk_d)]
+
+        # wishbone bus
+        self.bus = wishbone.Interface()
+        rd_ack = Signal()
+        wr_ack = Signal()
+        self.comb +=[
+            If(self.bus.we,
+               self.bus.ack.eq(wr_ack),
+            ).Else(
+                self.bus.ack.eq(rd_ack),
+            )
+        ]
+
+        # interrupts
+        self.submodules.ev = EventManager()
+        if hasattr(pads, 'rx'):
+            self.ev.rx_ready = EventSourcePulse()  # rising edge triggered, indicates FIFO is ready to read
```

Comment:
You can add a description to an EventSource object to get it documented, so you can turn the comment into documentation.

https://github.com/enjoy-digital/litex/blob/ae085782aed7f448cd205628532a7573781a34a6/litex/soc/interconnect/csr_eventmanager.py#L48

Reply:
just added comments to these, but left the # comment about rising/falling edge triggering because I can never remember which polarity is "Process" versus "Pulse". I don't know why they don't just use EventSourceRising(), EventSourceFalling(), and EventSourceLevel() as API names...